third_party/move/move-compiler-v2/src/pipeline/instruction_reordering.rs

```diff
@@ -0,0 +1,771 @@
+        let graph = &mut self.dd_graph.0;
+        let mut prepare_instrs: Vec<Bytecode> = vec![];
+        for (usage_offset, usage_instr) in self.instructions.iter().enumerate() {
+            let sources = usage_instr.sources();
+            if sources.len() < 2 {
+                // No need to insert `Prepare` for non-multi-source instructions.
+                continue;
+            }
+            let usage_offset = usage_offset as CodeOffset;
+            if let Some(defs) = graph.get_mut(&usage_offset) {
+                // We do not have to `Prepare` the last operand, as it can be brought
+                // to the top of the stack when actually needed by the use.
+                let without_last_len = if defs.is_empty() { 0 } else { defs.len() - 1 };
+                for (def, tmp) in defs.iter_mut().zip(sources.iter()).take(without_last_len) {
+                    if def.is_none() {
+                        // The definition is not explicit in the block.
+                        // So, let's insert a `Prepare` instruction.
+                        let prepare_offset =
+                            (self.original_block_len + prepare_instrs.len()) as CodeOffset;
+                        let prepare = Bytecode::Call(
+                            usage_instr.get_attr_id(),
+                            vec![],
+                            Operation::Prepare,
+                            vec![*tmp],
+                            None,
+                        );
+                        prepare_instrs.push(prepare);
+                        *def = Some(prepare_offset);
+                        prepare_use_map.insert(prepare_offset, UseInfo {
+                            offset: usage_offset,
+                            copy_only: false, // this may be updated later to true
+                        });
+                    }
+                }
+            }
+        }
+        self.instructions.extend(prepare_instrs);
+    }
+
+    /// Add all the constraints that restrict the reordering of instructions.
+    /// All these constraints must be respected when reordering instructions.
+    fn fill_constraints(&mut self) {
+        self.constraints = DependenceConstraints::compute_from(self);
+    }
+
+    /// Perform instruction reordering using the edge-ordered data dependence graph,
+    /// subject to the constraints.
+    fn reorder_instructions(&mut self) {
+        let mut reordered_block = vec![];
+        // Tracking instruction offsets that have already been included in the reordered block.
+        let mut taken = vec![false; self.instructions.len()];
+        let ref_args = self.get_ref_args();
+        let reads = self.get_reads();
+        let prepares = self.get_prepares();
+        // Start traversing from the bottom of the original block.
+        for (offset, instr) in self
+            .instructions
+            .iter()
+            .enumerate()
+            .take(self.original_block_len)
```

Comment:
The complexity of this iteration makes me think it would have been smarter to keep `Prepare` instructions in a separate list until this stage.  You don't really need to append them to the original basic block instructions.